src/core/container_async.rs

```diff
@@ -71,4 +72,53 @@ where
 
         log::debug!("Container {} is now ready!", self.id);
     }
+
+    pub async fn stop(&self) {
+        log::debug!("Stopping docker container {}", self.id);
+
+        self.docker_client.stop(&self.id).await
+    }
+
+    pub async fn start(&self) {
+        self.docker_client.start(&self.id).await
+    }
+
+    pub async fn rm(&self) {
+        log::debug!("Deleting docker container {}", self.id);
+
+        self.docker_client.rm(&self.id).await
+    }
+
+    #[allow(dead_code)]
+    async fn drop_async(&self) {
+        println!("start dropping");
+        let keep_container = var("KEEP_CONTAINERS")
+            .ok()
+            .and_then(|var| var.parse().ok())
+            .unwrap_or(false);
+
+        if keep_container {
+            self.stop().await
+        } else {
+            self.rm().await
+        }
+        println!("finished dropping");
+    }
+}
+
+// use futures::executor::block_on;
```

Comment:
I think that is a good call, we don't have any other option at this stage!

Reply:
:( unfortunately it doesn't work yet. It just stalled and never poll for the future underneath. I'll try more things to see if I can get a reliable Drop from an async test. I put more info on the comment above.

Journey continues.

Reply:
I wonder if that is the case because the whole test fn is a future and is being polled by another executor and we may stall the executor thread with this function.

Worst case users have to manually call `rm`/`stop` within async contexts for now, that would a bit sad though :(

Reply:
Ha! You are right, `#[tokio::test]` does treat the whole test as a future. `tokio_test::block_on` inside normal test on the other hand will work with a `Drop` method containing a small runtime of its own. But I don't like the feeling of restricting how the test should be written, if anything using `#[tokio::test]` is more favorable like you mentioned in another comment.  I'll keep digging. There might be other ways like spawning another thread and joining.

Reply:
After adding `#[tokio::test(threaded_scheduler)]` to test, `block_on` in Drop started doing the right thing and container is getting cleaned up!

I think it's a small price to pay to use the async API. I haven't read up much on threaded_scheduler and I think they just changed the API to `core_thread` and `max_thread`.

Reply:
If it works with `threaded_scheduler` then the initial assumption was right: `block_on` blocks the executor thread which means the future that actually does the `Container#rm` cannot make any progress because it is being executed within the same thread.

To make this easier to use, we might want to expose our own proc-macro at some point for this so we can do:

```rust
#[testcontainers::async_test]
async fn test() {

}
```

which would always use a threaded runtime.

I wonder if there could be any race-conditions where with bad luck, the future that does `Container#rm` still ends up being on the same thread as the one that we block and so things don't work _sometimes_. I am not sure if tokio is smart enough to move futures between executor threads if it noticed that one of them is blocked.

Reply:
Yeah, it does sound like a valid concern. Even thought Drop technically happens after all previous `.await` calls, there's no guaranteed the cleanup future gets pulled last. Let me clean it up a bit and try it on a slower VM and try limiting it to one core.

Reply:
Another way of trying this might be to just execute `cargo test` (for a single test) a thousand times or something :D
To increase the chances of it happening (if it ever happens), you also might want to consider setting the number of executor threads to 1 or possibly 2. I don't know if tokio also uses the current thread for executing futures or spawns new ones.

Reply:
Haha, that's exactly what I did and I ran it on my slower Linux VM. I didn't encounter any stalled and every run passed and all containers are removed. Tokio test API has newer control over this by allowing `core_thread` and `max_thread` which works as well. We might just be safer to implement `#[testcontainers::async_test]` trait, and fingers crossed tokio doesn't remove `threaded_scheduler`. The version I'm using didn't mention the macro at all. 

For now I'll start adding a more complicated reference implementation like redis to see if we are seeing other issue.

Reply:
> I didn't encounter any stalled and every run passed and all containers are removed.

That is good news! Thanks for testing!